core/src/main/java/com/linecorp/armeria/internal/logging/ContentPreviewerConfigurator.java

```diff
@@ -0,0 +1,129 @@
+        this.requestContentPreviewerFactory =
+                requireNonNull(requestContentPreviewerFactory, "requestContentPreviewerFactory");
+        this.responseContentPreviewerFactory =
+                requireNonNull(responseContentPreviewerFactory, "responseContentPreviewerFactory");
+    }
+
+    /**
+     * Sets up the request {@link ContentPreviewer} when {@code requestContentPreviewerFactory}
+     * is non-null and {@link ContentPreviewerFactory#get(RequestContext, HttpHeaders)} returns a
+     * {@link ContentPreviewer} instance that is not {@link ContentPreviewer#isDisabled()}.
+     */
+    public HttpRequest maybeSetUpRequestContentPreviewer(RequestContext ctx, HttpRequest req) {
+        requireNonNull(ctx, "ctx");
+        requireNonNull(req, "req");
+        if (requestContentPreviewerFactory == null) {
+            return req;
+        }
+
+        final RequestHeaders headers = req.headers();
+        final ContentPreviewer requestContentPreviewer = requestContentPreviewerFactory.get(ctx, headers);
+        if (requestContentPreviewer.isDisabled()) {
+            return req;
+        }
+
+        ctx.logBuilder().requestContentPreviewer(requestContentPreviewer);
+        requestContentPreviewer.onHeaders(headers);
+        return new FilteredHttpRequest(req) {
+            @Override
+            protected HttpObject filter(HttpObject obj) {
+                if (obj instanceof HttpData) {
+                    requestContentPreviewer.onData((HttpData) obj);
+                }
+                return obj;
+            }
+        };
+    }
+
+    /**
+     * Sets up the response {@link ContentPreviewer} when {@code responseContentPreviewerFactory}
+     * is non-null.
+     */
+    public HttpResponse maybeSetUpResponseContentPreviewer(RequestContext ctx, HttpResponse res) {
+        requireNonNull(ctx, "ctx");
+        requireNonNull(res, "res");
+        if (responseContentPreviewerFactory == null) {
+            return res;
+        }
+
+        return new FilteredHttpResponse(res) {
+            @Nullable
+            ContentPreviewer responseContentPreviewer;
+
+            @Override
+            protected HttpObject filter(HttpObject obj) {
+                if (obj instanceof ResponseHeaders) {
+                    final ResponseHeaders headers = (ResponseHeaders) obj;
+
+                    // Skip informational headers.
+                    final String status = headers.get(HttpHeaderNames.STATUS);
+                    if (ArmeriaHttpUtil.isInformational(status)) {
+                        return obj;
+                    }
+
+                    responseContentPreviewer = responseContentPreviewerFactory.get(ctx, headers);
+                    ctx.logBuilder().responseContentPreviewer(responseContentPreviewer);
+                    responseContentPreviewer.onHeaders(headers);
+                } else if (obj instanceof HttpData) {
+                    assert responseContentPreviewer != null;
```

Comment:
Do we have some validation to make sure `ResponseHeaders` are first published object? I wonder if we are replacing a downstream HTTP protocol exception with a NPE here. I think it doesn't hurt much to just add the null check here to not worry.